### PR TITLE
More specific browser detection

### PR DESCRIPTION
**Pull request by mightyiam**

### CODE REVIEW

**Comment by mightyiam** [comment]
In #624 it seems that more specific browser detection is required.

Specifically, for detecting Chrome, Safari and WebKit, instead of treating them all as WebKit.

File: Gruntfile.js [hunk 1]
```diff
@@ -254,6 +254,8 @@ module.exports = function (grunt) {
                         "document-structure-manager.js",
                     "<%= yeoman.app %>/wymeditor/editor/gecko.js",
                     "<%= yeoman.app %>/wymeditor/editor/webkit.js",
+                    "<%= yeoman.app %>/wymeditor/editor/blink.js",
+                    "<%= yeoman.app %>/wymeditor/editor/safari.js",
```
Comment (mightyiam) [inline Gruntfile.js]: Concatenation for distribution

File: bower.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
   "devDependencies": {
     "requirejs": "~2.1",
     "jquery": "~1.4.4",
-    "jquery.browser": "~0.0.6",
+    "jquery.browser": "~0.0.7",
```
Comment (mightyiam) [inline bower.json]: Officially newer version while we're at it.

File: src/test/load-src.dev.js [hunk 3]
```diff
@@ -25,6 +25,8 @@ function loadWymSrc(srcPath, extraRequirements, jqueryVersion) {
             srcPath + 'wymeditor/editor/trident-pre-7.js',
             srcPath + 'wymeditor/editor/trident-7.js',
             srcPath + 'wymeditor/editor/webkit.js',
+            srcPath + 'wymeditor/editor/blink.js',
+            srcPath + 'wymeditor/editor/safari.js',
```
Comment (mightyiam) [inline src/test/load-src.dev.js]: Load them in dev–env.

File: src/wymeditor/core.js [hunk 4]
```diff
@@ -1140,3 +1140,39 @@ WYMeditor.Helper = {
...
+
+/**
+    WYMeditor._getWymClass
+    ======================
+
+    Returns the constructor for the browser-specific wymeditor instance.
+
+    If does not detect a supported browser, returns false;
+*/
+WYMeditor._getWymClass = function () {
```
Comment (mightyiam) [inline src/wymeditor/core.js]: The new browser detecting logic.

File: src/wymeditor/editor/base.js [hunk 5]
```diff
@@ -33,29 +34,17 @@ WYMeditor.editor.prototype._init = function () {
...
-            // https://stackoverflow.com/questions/20655470
-            //WymClass = new WYMeditor.WymClassBlink(wym);
-            // For now we use the WebKit editor class in Blink.
-            WymClass = new WYMeditor.WymClassWebKit(wym);
-        } else {
-            WymClass = new WYMeditor.WymClassWebKit(wym);
-        }
-    }
+    // Get the constructor for the browser-specific instance
+    WymClass = WYMeditor._getWymClass();
```
Comment (mightyiam) [inline src/wymeditor/editor/base.js]: Delegate to the new `WYMeditor._getWymClass`.

Comment (mightyiam) [inline src/wymeditor/editor/base.js]: Also, notice that the constructor itself gets assigned to `WymClass` (justifies the capital W).

Only below it gets instantiated and assigned to the newly added `browserInstance`.

File: src/wymeditor/editor/base.js [hunk 6]
```diff
@@ -33,29 +34,17 @@ WYMeditor.editor.prototype._init = function () {
...
 
-    if (WymClass === false) {
+    if (!WymClass) {
+        // We don't support this browser. Don't initialize.
```
Comment (winhamwr) [inline src/wymeditor/editor/base.js]: It feels like that if we can't reliably print to the console via `console.warn` to give a developer some feedback, we should throw an exception. This silent failure would be really frustrating behavior for someone to debug.

File: src/wymeditor/editor/base.js (hunk 5 above)
Comment (winhamwr) [inline src/wymeditor/editor/base.js]: :+1: 
Good level of encapsulation.

For a name, how about `_getWymClassForBrowser` to make it clear that the class logic is about the user's web browser? Then, you don't actually need the comment. Better names > comments

File: src/wymeditor/core.js [hunk 7]
```diff
@@ -1140,3 +1140,39 @@ WYMeditor.Helper = {
...
+        case "mozilla":
+            return WYMeditor.WymClassGecko;
+        case "chrome":
+            return WYMeditor.WymClassBlink;
+        case "safari":
+            return WYMeditor.WymClassSafari;
+    }
+
+    if (jQuery.browser.webkit) {
+        return WYMeditor.WymClassWebKit;
```
Comment (winhamwr) [inline src/wymeditor/core.js]: What case does this handle?

Comment (mightyiam) [inline src/wymeditor/core.js]: 1. https://en.wikipedia.org/wiki/Midori_%28web_browser%29
1. https://en.wikipedia.org/wiki/Arora_%28web_browser%29

Whatever else anyone may come up with.

Comment (winhamwr) [inline src/wymeditor/core.js]: Makes sense.

File: src/wymeditor/core.js [hunk 8]
```diff
@@ -1140,3 +1140,39 @@ WYMeditor.Helper = {
...
+
+/**
+    WYMeditor._getWymClassForBrowser
+    ================================
+
+    Returns the constructor for the browser-specific wymeditor instance.
+
+    If does not detect a supported browser, returns false;
+*/
+WYMeditor._getWymClassForBrowser = function () {
```
Comment (mightyiam) [inline src/wymeditor/core.js]: @winhamwr here's your suggested name.

Comment (winhamwr) [inline src/wymeditor/core.js]: :+1:

File: src/wymeditor/core.js [hunk 9]
```diff
@@ -1140,3 +1140,41 @@ WYMeditor.Helper = {
...
+            return WYMeditor.WymClassBlink;
+        case "safari":
+            return WYMeditor.WymClassSafari;
+    }
+
+    if (jQuery.browser.webkit) {
+        return WYMeditor.WymClassWebKit;
+    }
+
+    WYMeditor.console.warn("WYMeditor could not instantiate: this browser " +
```
Comment (mightyiam) [inline src/wymeditor/core.js]: Browser not supported warning @winhamwr.

**Comment by winhamwr** [comment]
Looks good!

**Comment by mightyiam** [comment]
Thank you!

**Pull request merged**